discord/discord_video_source.cc

```diff
@@ -0,0 +1,255 @@
+          std::move(buffer_handle), size, gfx::BufferFormat::RGBA_8888,
+          gfx::BufferUsage::GPU_READ, base::DoNothing());
+  auto mailbox = sii->CreateSharedImage(
+      buffer_impl.get(), gpu_factories_->GpuMemoryBufferManager(),
+      gfx::ColorSpace::CreateSRGB(), kTopLeft_GrSurfaceOrigin,
+      kOpaque_SkAlphaType,
+      gpu::SHARED_IMAGE_USAGE_GLES2 | gpu::SHARED_IMAGE_USAGE_RASTER |
+          gpu::SHARED_IMAGE_USAGE_DISPLAY | gpu::SHARED_IMAGE_USAGE_SCANOUT);
+  gpu::MailboxHolder mailbox_holder_array[media::VideoFrame::kMaxPlanes];
+  mailbox_holder_array[0] = gpu::MailboxHolder(
+      mailbox, sii->GenUnverifiedSyncToken(),
+      gpu_factories_->ImageTextureTarget(buffer_impl->GetFormat()));
+
+  scoped_refptr<media::VideoFrame> video_frame =
+      media::VideoFrame::WrapExternalGpuMemoryBuffer(
+          gfx::Rect(size), size, std::move(buffer_impl), mailbox_holder_array,
+          base::DoNothing(), base::TimeDelta::FromMicroseconds(timestamp_us));
+  PostCrossThreadTask(
+      *io_task_runner_.get(), FROM_HERE,
+      CrossThreadBindOnce(&VideoSourceDelegate::DoRenderFrameOnIOThread,
+                          WrapRefCounted(this), video_frame,
+                          base::TimeTicks()));
+  video_frame->AddDestructionObserver(ConvertToBaseOnceCallback(
+      CrossThreadBindOnce(releaseCB, CrossThreadUnretained(userData))));
+}
+#endif
+
+MediaStreamDiscordVideoSource::MediaStreamDiscordVideoSource(
+    const std::string& streamId)
+    : stream_id_(streamId) {}
+
+MediaStreamDiscordVideoSource::~MediaStreamDiscordVideoSource() {
+  StopSourceImpl();
+}
+
+void MediaStreamDiscordVideoSource::StartSourceImpl(
+    VideoCaptureDeliverFrameCB frame_callback,
+    EncodedVideoFrameCB encoded_frame_callback) {
+  delegate_ = base::MakeRefCounted<VideoSourceDelegate>(
+      io_task_runner(), std::move(frame_callback));
+  {
+    std::lock_guard<std::mutex> lock(stream_mutex_);
+    if (!by_stream_id_) {
+      by_stream_id_ =
+          new std::unordered_map<std::string,
+                                 scoped_refptr<VideoSourceDelegate>>();
+    }
+    by_stream_id_->erase(stream_id_);
+    by_stream_id_->emplace(stream_id_, delegate_);
+  }
+
+  OnStartDone(mojom::MediaStreamRequestResult::OK);
+}
+
+void MediaStreamDiscordVideoSource::StopSourceImpl() {
+  if (delegate_.get()) {
+    std::lock_guard<std::mutex> lock(stream_mutex_);
+    auto it = by_stream_id_->find(stream_id_);
+    if (it != by_stream_id_->end() && it->second == delegate_) {
+      by_stream_id_->erase(it);
+    }
+    delegate_.reset();
+  }
+}
+
+void MediaStreamDiscordVideoSource::VideoSourceDelegate::OnFrame(
+    const std::string& streamId,
+    const DiscordFrame& frame,
+    DiscordFrameReleaseCB releaseCB,
+    void* userData) {
+  gfx::Size size(frame.width, frame.height);
+  scoped_refptr<media::VideoFrame> video_frame;
+  auto type = static_cast<webrtc::VideoFrameBuffer::Type>(frame.type);
```

Comment:
also, this maybe super nitty but... I kinda favor defining our own type enum for this api if it's not too much hassle. It's conceivable that it may diverge e.g. when we do media framework integration with our external webrtc

Reply:
Not too much hassle at all